src/cpp/ext/gsm/metadata_exchange.cc

```diff
@@ -0,0 +1,243 @@
+    "canonical_service";
+// The keys that will be used for the peer attributes when recording metrics.
+constexpr absl::string_view kPeerTypeAttribute = "gsm.remote_workload_type";
+constexpr absl::string_view kPeerPodNameAttribute =
+    "gsm.remote_workload_pod_name";
+constexpr absl::string_view kPeerContainerNameAttribute =
+    "gsm.remote_workload_container_name";
+constexpr absl::string_view kPeerNamespaceNameAttribute =
+    "gsm.remote_workload_namespace_name";
+constexpr absl::string_view kPeerClusterNameAttribute =
+    "gsm.remote_workload_cluster_name";
+constexpr absl::string_view kPeerLocationAttribute =
+    "gsm.remote_workload_location";
+constexpr absl::string_view kPeerProjectIdAttribute =
+    "gsm.remote_workload_project_id";
+constexpr absl::string_view kPeerCanonicalServiceAttribute =
+    "gsm.remote_workload_canonical_service";
+
+upb_StringView AbslStrToUpbStr(absl::string_view str) {
+  return upb_StringView_FromDataAndSize(str.data(), str.size());
+}
+
+absl::string_view UpbStrToAbslStr(upb_StringView str) {
+  return absl::string_view(str.data, str.size);
+}
+
+void AddStringKeyValueToStructProto(google_protobuf_Struct* struct_pb,
+                                    absl::string_view key,
+                                    absl::string_view value, upb_Arena* arena) {
+  google_protobuf_Value* value_pb = google_protobuf_Value_new(arena);
+  google_protobuf_Value_set_string_value(value_pb, AbslStrToUpbStr(value));
+  google_protobuf_Struct_fields_set(struct_pb, AbslStrToUpbStr(key), value_pb,
+                                    arena);
+}
+
+absl::string_view GetStringValueFromAttributeMap(
+    const opentelemetry::sdk::common::AttributeMap& map,
+    absl::string_view key) {
+  const auto& attributes = map.GetAttributes();
+  const auto it = attributes.find(std::string(key));
+  if (it == attributes.end()) {
+    return "unknown";
+  }
+  const auto* string_value = absl::get_if<std::string>(&it->second);
+  if (string_value == nullptr) {
+    return "unknown";
+  }
+  return *string_value;
+}
+
+absl::string_view GetStringValueFromUpbStruct(google_protobuf_Struct* struct_pb,
+                                              absl::string_view key,
+                                              upb_Arena* arena) {
+  if (struct_pb == nullptr) {
+    return "unknown";
+  }
+  google_protobuf_Value* value_pb = google_protobuf_Value_new(arena);
+  bool present = google_protobuf_Struct_fields_get(
+      struct_pb, AbslStrToUpbStr(key), &value_pb);
+  if (present) {
+    if (google_protobuf_Value_has_string_value(value_pb)) {
+      return UpbStrToAbslStr(google_protobuf_Value_string_value(value_pb));
+    }
+  }
+  return "unknown";
+}
+
+}  // namespace
+
+ServiceMeshLabelsInjector::ServiceMeshLabelsInjector(
+    const opentelemetry::sdk::common::AttributeMap& map) {
+  upb::Arena arena;
+  auto* metadata = google_protobuf_Struct_new(arena.ptr());
+  // Assume kubernetes for now
```

Comment:
Even if we're not supporting the labels for other environments right now, I think we should at least be smart enough not to try to populate the GKE labels if the environment is not GKE.  If we see a non-GKE environment, we should populate only the remote_workload_type and nothing else for now.  Then we can easily add support for other types later, but at least we won't be populating the wrong labels if someone runs on such an environment with this code.

Reply:
Done